### ISSUE TITLE
Interactive mode for Mistral example

### ISSUE THREAD

**Issue by evgenyigumnov** [huggingface/candle#1896]
Hello

How to create interactive mode for  https://github.com/huggingface/candle/tree/main/candle-examples/examples/mistral ?

I try to use this interactive logic from:

https://github.com/huggingface/candle/blob/main/candle-examples/examples/quantized/main.rs

But it is not working. When I try model.forward for next promt it is not work.

I have error:

Error: shape mismatch in broadcast_add, lhs: [1, 32, 177, 353], rhs: [1, 1, 177, 177]

```

                let context_size = if index > 0 { 1 } else { tokens.len() };
                let start_pos = tokens.len().saturating_sub(context_size);
                let ctxt = &tokens[start_pos..];
                let input = Tensor::new(ctxt, &self.device)?.unsqueeze(0)?;
                let logits = match &mut self.model {
                    Model::Mistral(m) => m.forward(&input, start_pos)?,  <<<<< !!!
                    Model::Quantized(m) => m.forward(&input, start_pos)?,   <<<<  !!!
                };
```

**Comment by evgenyigumnov** [comment]
Unfortanly all **quantized**  models absolutly stupid for my task.

I need mistral without  **quantized**

```
[package]
name = "ai-server"
version = "0.1.0"
edition = "2021"

[dependencies]

candle-nn = "0.4.1"
candle-core = "0.4.1"
candle-datasets = "0.4.1"
candle-transformers = "0.4.1"
candle-examples = "0.4.1"
hf-hub = "0.3.2"
tokenizers = "0.15.2"
anyhow = "1.0.81"
clap = { version = "4.5.3", features = ["derive"] }
tracing = "0.1.37"
tracing-chrome = "0.7.1"
tracing-subscriber = "0.3.7"

[build-dependencies]
bindgen_cuda = { version = "0.1.1", optional = true }


[features]
cuda = ["candle-core/cuda", "candle-nn/cuda", "candle-transformers/cuda", "dep:bindgen_cuda"]

```




```
use anyhow::{Error as E, Result};
use clap::Parser;

use candle_transformers::models::mistral::{Config, Model as Mistral};

use candle_core::{DType, Device, Tensor};
use candle_examples::token_output_stream::TokenOutputStream;
use candle_nn::VarBuilder;
use candle_transformers::generation::LogitsProcessor;
use hf_hub::{api::sync::Api, Repo, RepoType};
use tokenizers::Tokenizer;


struct TextGeneration {
    model: Mistral,
    device: Device,
    tokenizer: TokenOutputStream,
    logits_processor: LogitsProcessor,
    repeat_penalty: f32,
    repeat_last_n: usize,
}

impl TextGeneration {
    #[allow(clippy::too_many_arguments)]
    fn new(
        model: Mistral,
        tokenizer: Tokenizer,
        seed: u64,
        temp: Option<f64>,
        top_p: Option<f64>,
        repeat_penalty: f32,
        repeat_last_n: usize,
        device: &Device,
    ) -> Self {
        let logits_processor = LogitsProcessor::new(seed, temp, top_p);
        Self {
            model,
            tokenizer: TokenOutputStream::new(tokenizer),
            logits_processor,
            repeat_penalty,
            repeat_last_n,
            device: device.clone(),
        }
    }

    fn run(&mut self, prompts: Vec<&str>, sample_len: usize) -> Result<()> {
        use std::io::Write;
        self.tokenizer.clear();
        for prompt in prompts {
            let mut tokens = self
                .tokenizer
                .tokenizer()
                .encode(prompt, true)
                .map_err(E::msg)?
                .get_ids()
                .to_vec();
            println!("{}", prompt);
            std::io::stdout().flush()?;

            let mut generated_tokens = 0usize;
            let eos_token = match self.tokenizer.get_token("</s>") {
                Some(token) => token,
                None => anyhow::bail!("cannot find the </s> token"),
            };
            let start_gen = std::time::Instant::now();
            for index in 0..sample_len {
                let context_size = if index > 0 { 1 } else { tokens.len() };
                let start_pos = tokens.len().saturating_sub(context_size);
                let ctxt = &tokens[start_pos..];
                let input = Tensor::new(ctxt, &self.device)?.unsqueeze(0)?;
                let logits = self.model.forward(&input, start_pos)?;
                let logits = logits.squeeze(0)?.squeeze(0)?.to_dtype(DType::F32)?;
                let logits = if self.repeat_penalty == 1. {
                    logits
                } else {
                    let start_at = tokens.len().saturating_sub(self.repeat_last_n);
                    candle_transformers::utils::apply_repeat_penalty(
                        &logits,
                        self.repeat_penalty,
                        &tokens[start_at..],
                    )?
                };

                let next_token = self.logits_processor.sample(&logits)?;
                tokens.push(next_token);
                generated_tokens += 1;
                if next_token == eos_token {
                    break;
                }
                if let Some(t) = self.tokenizer.next_token(next_token)? {
                    print!("{t}");
                    std::io::stdout().flush()?;
                }
            }
            let dt = start_gen.elapsed();
            if let Some(rest) = self.tokenizer.decode_rest().map_err(E::msg)? {
                print!("{rest}");
            }
            std::io::stdout().flush()?;
            println!(
                "\n{generated_tokens} tokens generated ({:.2} token/s)",
                generated_tokens as f64 / dt.as_secs_f64(),
            );


        }

        Ok(())
    }
}

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// Run on CPU rather than on GPU.
    #[arg(long)]
    cpu: bool,

    /// Enable tracing (generates a trace-timestamp.json file).
    #[arg(long)]
    tracing: bool,

    #[arg(long)]
    use_flash_attn: bool,


    /// The temperature used to generate samples.
    #[arg(long)]
    temperature: Option<f64>,

    /// Nucleus sampling probability cutoff.
    #[arg(long)]
    top_p: Option<f64>,

    /// The seed to use when generating random samples.
    #[arg(long, default_value_t = 299792458)]
    seed: u64,



    #[arg(long)]
    model_id: Option<String>,

    #[arg(long, default_value = "main")]
    revision: String,

    #[arg(long)]
    tokenizer_file: Option<String>,

    #[arg(long)]
    weight_files: Option<String>,


    /// Penalty to be applied for repeating tokens, 1. means no penalty.
    #[arg(long, default_value_t = 1.1)]
    repeat_penalty: f32,

    /// The context size to consider for the repeat penalty.
    #[arg(long, default_value_t = 64)]
    repeat_last_n: usize,
}

fn main() -> Result<()> {
    use tracing_chrome::ChromeLayerBuilder;
    use tracing_subscriber::prelude::*;

    let args = Args::parse();
    let _guard = if args.tracing {
        let (chrome_layer, guard) = ChromeLayerBuilder::new().build();
        tracing_subscriber::registry().with(chrome_layer).init();
        Some(guard)
    } else {
        None
    };
    println!(
        "avx: {}, neon: {}, simd128: {}, f16c: {}",
        candle_core::utils::with_avx(),
        candle_core::utils::with_neon(),
        candle_core::utils::with_simd128(),
        candle_core::utils::with_f16c()
    );
    println!(
        "temp: {:.2} repeat-penalty: {:.2} repeat-last-n: {}",
        args.temperature.unwrap_or(0.),
        args.repeat_penalty,
        args.repeat_last_n
    );

    let start = std::time::Instant::now();
    let api = Api::new()?;
    let model_id = match args.model_id {
        Some(model_id) => model_id,
        None => {
            "mistralai/Mistral-7B-v0.1".to_string()
        }
    };
    let repo = api.repo(Repo::with_revision(
        model_id,
        RepoType::Model,
        args.revision,
    ));
    let tokenizer_filename = match args.tokenizer_file {
        Some(file) => std::path::PathBuf::from(file),
        None => repo.get("tokenizer.json")?,
    };
    let filenames = match args.weight_files {
        Some(files) => files
            .split(',')
            .map(std::path::PathBuf::from)
            .collect::<Vec<_>>(),
        None => {
                candle_examples::hub_load_safetensors(&repo, "model.safetensors.index.json")?
        }
    };
    println!("retrieved the files in {:?}", start.elapsed());
    let tokenizer = Tokenizer::from_file(tokenizer_filename).map_err(E::msg)?;

    let start = std::time::Instant::now();
    let config = Config::config_7b_v0_1(args.use_flash_attn);
    let device = candle_examples::device(args.cpu)?;

    let (model, device) = {
        let dtype = if device.is_cuda() {
            DType::BF16
        } else {
            DType::F32
        };
        let vb = unsafe { VarBuilder::from_mmaped_safetensors(&filenames, dtype, &device)? };
        let model = Mistral::new(&config, vb)?;
        (model, device)
    };

    println!("loaded the model in {:?}", start.elapsed());
    let mut pipeline = TextGeneration::new(
        model,
        tokenizer,
        299792458,
        Some(0.0),
        Some(00.0),
        1.1,
        64,
        &device,
    );

    let prompt1 = r#"
<s>
### USER ###

[Document 1]

Invoice 234

Tomato 3 kg 400$
Nuts 4 kg 1600$

Total: 2000$

[Document 2]

Medical card 769
Client name: John Smith
History of illness: Bla Bla Bla
Last visit hospital date: 17 march 1979
Personal doctor name: Anna Zuzuko

Compare Document 1 and Document 2. These documents in the same field or subject? Give short answer like: Yes/No

### ASSISTANCE ###
Answer:
"#;

    let prompt2 = r#"
<s>
### USER ###

[Document 1]

Invoice 234

Tomato 3 kg 400$
Nuts 4 kg 1600$

Total: 2000$

[Document 2]

Invoice 257

Meat 10 kg 100$

Total: 1000$


Compare Document 1 and Document 2. These documents in the same field or subject? Give short answer like: Yes/No

### ASSISTANCE ###
Answer:
"#;

    let prompts  = vec![prompt1, prompt2];
    pipeline.run(prompts, 5)?;
    Ok(())
}
```




```
root@C.10177988:~/ai$ cargo run --features cuda --release --
    Finished release [optimized] target(s) in 0.28s
     Running `target/release/ai-server`
avx: false, neon: false, simd128: false, f16c: false
temp: 0.00 repeat-penalty: 1.10 repeat-last-n: 64
retrieved the files in 27.942098ms
loaded the model in 20.814441096s

<s>
### USER ###

[Document 1]

Invoice 234

Tomato 3 kg 400$
Nuts 4 kg 1600$

Total: 2000$

[Document 2]

Medical card 769
Client name: John Smith
History of illness: Bla Bla Bla
Last visit hospital date: 17 march 1979
Personal doctor name: Anna Zuzuko

Compare Document 1 and Document 2. These documents in the same field or subject? Give short answer like: Yes/No

### ASSISTANCE ###
Answer:


Yes, these documents
5 tokens generated (4.40 token/s)

<s>
### USER ###

[Document 1]

Invoice 234

Tomato 3 kg 400$
Nuts 4 kg 1600$

Total: 2000$

[Document 2]

Invoice 257

Meat 10 kg 100$

Total: 1000$


Compare Document 1 and Document 2. These documents in the same field or subject? Give short answer like: Yes/No

### ASSISTANCE ###
Answer:

Error: shape mismatch in broadcast_add, lhs: [1, 32, 138, 300], rhs: [1, 1, 138, 138]

```

**Comment by evgenyigumnov** [comment]
I check my task on Q RTX 8000  Graphics Card   45 Gb GPU  price: 4.000$  with **quantized** mixtral-8x7b-v0.1.Q4_K_M.gguf

Result is terrible. After **quantized** proccess models start be very stupid. I need mistal without  **quantized**.
I need interactive mode with CUDA and with split prompt:
```
    /// Process prompt elements separately.
    #[arg(long)]
    split_prompt: bool,
```

**Comment by groovybits** [comment]
I find the quantized mistral model does the same for me (at least in the mistral example), rarely produce sane output. In fact the mistral --quantized one will say awful things at times that has no context to what I put in as a prompt :D it's scary what it can say :/ I'm not sure what I am doing wrong there.

I see a similar errors of shape mismatch the regular mistral example and gemma when I loop around them, same thing happens, when I am continuously looping around the code where if I loop around it even with it in a separate function initializing each time, eventually it hits a point where it outputs those type of misshape messages. In llama.cpp the quantized models do not behave the same, are great, I'm still wondering why that one mistral quantized one is so much different?

When I retry the loop and retry when bad weights are produced mine also seems to hit the model forward shape issue, initially it gets all weights zero but if I retry again at the forward part it does the same broadcast misshapen issue.

I am very interested in how to loop the model forever without these issues. The zero weights happens to me even if I can loop it, eventually it fails.

Here is what I have from taking the example and trying to do it in an easy separate crate like usage of my program:
https://github.com/groovybits/rsllm/blob/main/src/candle_gemma.rs#L95

I am retrying there when the weights are zero, happens after 7-13+ loops around the main gemma() function and my mistral one is here which they both have the same issue.
https://github.com/groovybits/rsllm/blob/main/src/candle_mistral.rs#L113

I was going to file a ticket to ask more about this, am still digging into it more to understand it. The stable diffusion model does this to me too if I give it too long of input sentence. Often it feels I am not clearing the above mistral/candle models somehow even though I don't think any parts are staying initialized between loops. 

So I wonder if this is related and something when looping causes this breaking of it functioning till exiting/restarting the application w/out being able to re-init each loop over using the model? It sounds like the quantized example is the one I should have started my code's candle implementations with :/

**Comment by EricLBuehler** [comment]
During my work with quantized mistral models, this strange behavior (no error, but bad results) can be caused by reshape operations that end up shuffling data or by the use of GPT-J vs GPT-NEOX style RoPE interleave.

If you are interested in inference of quantized Mistral, perhaps you can check out [mistral.rs](https://github.com/EricLBuehler/mistral.rs). It provides an OpenAI API server with Quantized Mistral and other models and is fast with an optimized backend.

**Comment by groovybits** [comment]
Yes I looked through it and am impressed! The output of the quantized mistral today seemed like it would get me wached if I let it out :D like it talks about all this illegal stuff, it's terrible and I suspect needs to be updated to not use that one and do it right or remove it. I mean it's really weird and wrong what it can say without any provocation, the more complex the prompt the more you should expect do be aghast at what it says back!

**Comment by LaurentMazare** [comment]
Could you provide some examples of prompt that trigger these weird behaviors, and also detail which model you've been using (instruct vs not instruct etc)? If you can give a way to reproduce the issue, happy to investigate it.

**Comment by evgenyigumnov** [comment]
> i

Thank you for promotion of project. But unfortanly you are using only quantized Mistral. I need original Mistral...

**Comment by evgenyigumnov** [comment]
@LaurentMazare Hello, I hope this message finds you well. I wanted to follow up on a code example I previously provided, aimed at reproducing a certain issue we've encountered. My supervisor has been inquiring about the progress on this matter daily, and I was hoping to seek your expertise. Could you kindly assist us in resolving this issue? Your support would be greatly appreciated.

**Comment by LaurentMazare** [comment]
I think you just forgot to flush the kv cache between processing the two documents. You can do so with the following diff.
```diff
diff --git a/src/main.rs b/src/main.rs
index 9f29a95..d19044f 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -61,6 +61,7 @@ impl TextGeneration {
                 Some(token) => token,
                 None => anyhow::bail!("cannot find the </s> token"),
             };
+            self.model.clear_kv_cache();
             let start_gen = std::time::Instant::now();
             for index in 0..sample_len {
                 let context_size = if index > 0 { 1 } else { tokens.len() };
```

**Comment by evgenyigumnov** [comment]
@LaurentMazare 

it worked! Thanks a lot! If you are in Astana, write to me in telegram ( account: ievkz ) ! I'll treat you to beer!
